### ISSUE TITLE
Add support for date/time constructs

### ISSUE THREAD

**Issue by holograph** [wix/accord#90]
Predominantly `java.time.*`, `java.util.Date` and possibly (in a separate project) JODA.
Example combinators might be `p.birthDate is before(Date.now().minus(18, TimeUnit.YEARS))`, `before` etc.

**Comment by zbeckman** [comment]
Big, big +1 on this. Dates a always popping up... Having basic support that includes the following tests would likely cover the bases:

- before
- after
- equalTo
- between

To be very fancy, you could add a:

- within

The idea of `within` is to give some degree of fuzziness. For instance, "within 1 hour" or "within 1 day." Probably just syntactic sugar for `between`.

Finally I'd say focus on Java date types. While I love Joda, it is now deprecated and will probably start to fade away... we don't use it anymore here. (Just some older legacy code).

**Comment by holograph** [comment]
There's some setbacks with `within` because of the way overload resolution works in Scala. The core Accord DSL (`import com.wix.accord.dsl._`) already includes a `within`, but importing a different overload via `import com.wix.accord.java8._` results in ambiguous reference errors:

```
[error] /Users/tomerga/dev/accord/java8/src/test/scala/com/wix/accord/java8/tests/TemporalOpsTests.scala:62: reference to within is ambiguous;
[error] it is imported twice in the same scope by
[error] import com.wix.accord.java8._
[error] and import com.wix.accord.dsl._
[error]   val withinTimeUnitValidator = tomorrow is within( 1L, ChronoUnit.WEEKS ) of now
```

Looks like we'll have to forego `within` or go about it a different way. In either case, this obviously isn't important enough to delay 0.6.1, so I'm pushing it back to the 0.7 milestone.

You can track progress on in the [issue90 branch](https://github.com/wix/accord/tree/issue90).

**Comment by holograph** [comment]
Alright, managed to work around the issue and ran smack into another. Fundamentally, Scala postfix expressions work left-to-right naïvely; in effect, `p.birthDate is within( someDuration ) of someDate` would translate to: `p.birthDate.is(within(someDuration)).of.someDate`.

There are really two solutions here:
1. Require users to instead use `is within( someDuration ).of( someDate )`. This is inconsistent with the existing APIs and design goals of the DSL, and is also impossible to enforce without obtuse error messages.
2. Put in infrastructure to allow validator "builders" (a reverse-associativity version of `is`). This may well be more work than is worthwhile for this particular feature.

@zbeckman, can you weigh in on what you would consider the best syntax for your use cases?

**Comment by zbeckman** [comment]
Aigh. Well... functionality probably wins out over lack of functionality. But I feel your pain on the syntax. This is ultimately a "nice to have" and can be emulated using `(is after X) and (is before Y)` so... aigh. Up to you. :-) I do like have the simpler functionality even though it flies in the face of the DSL, so if you don't find it too offensive I'd probably go ahead with `is within(X).of(Y)`.

**Comment by holograph** [comment]
(Pinging @zbeckman)

There's a bunch of functionality implemented in the [issue90 branch](https://github.com/wix/accord/tree/issue90). I'm not an altogether heavy user of either library, so I probably got some of the basics wrong; in particular I'm not exactly happy about the way formatting is handled. Here's what it looks like for joda-time:

```scala
  val now: Instant = DateTime.now().toInstant
  val epoch: ReadableInstant = DateTime.parse( "1970-01-01T00:00:00Z" )
  val otherEpoch = new Instant( 0L )
  val lastWeek: ReadableInstant = now.minus( Duration.standardDays( 7 ) )
  val nextWeek: ReadableInstant = now.plus( Duration.standardDays( 7 ) )
  val lastYear: ReadableInstant = now.minus( Duration.standardDays( 365 ) )
  val tomorrow: ReadableInstant = now.plus( Duration.standardDays( 1 ) )

  val atEpoch = validator[ ReadableInstant ] { t => t is equalTo( epoch ) }
  val notAtEpoch = validator[ ReadableInstant ] { t => t is notEqualTo( epoch ) }
  val duringLastYear = validator[ ReadableInstant ] { t => t is between( lastYear, now ) }

  atEpoch( otherEpoch ) should be( aSuccess )
  atEpoch( now ) should be( aFailure )
  notAtEpoch( epoch ) should be( aFailure )
  notAtEpoch( now ) should be( aSuccess )
  duringLastYear( epoch ) should be( aFailure )
  duringLastYear( lastWeek ) should be( aSuccess )
  duringLastYear( nextWeek ) should be( aFailure )
```

If you can have a look at the DSL tests ([Java 8](https://github.com/wix/accord/blob/issue90/java8/src/test/scala/com/wix/accord/java8/tests/TemporalOpsTests.scala), [Joda](https://github.com/wix/accord/blob/issue90/joda/src/test/scala/com/wix/accord/joda/tests/ReadableInstantOpsTests.scala)) some feedback on this would be greatly appreciated!

**Comment by holograph** [comment]
All changes and features are now available in the master branch (and with 0.7-SNAPSHOT). I would appreciate any feedback!

**Comment by holograph** [comment]
No feedback means no more enhancements :-) Merged and will be released with 0.7 very soon.
